predictive_test_selection/data_collector.py

```diff
@@ -0,0 +1,236 @@
+
+        for dev_branch in DEV_BRANCHES:
+            cmd = ["git", "diff", f"{dev_branch}...{current_branch}", "--name-only"]
+            changed_files = [line.strip() for line in subprocess.check_output(cmd).decode().split("\n") if line.strip()]
+
+            current_cardinality = len(changed_files)
+
+            if current_cardinality < max_cardinality:
+                max_cardinality = current_cardinality
+                self.dev_branch = dev_branch
+                self.changed_files = changed_files
+                self.file_cardinality = current_cardinality
+
+            if max_cardinality == 0:
+                raise AbortDataCollection(f"the code change is aligned with dev branch {dev_branch}.")
+
+    def _count_modifications(self) -> None:
+        """
+        Counts the number of modifications that have been made to the files involved in this change for each interval (in days)
+        specified in the past CHANGE_LOOK_BACK parameter.
+        """
+        modification_count: List[int] = []
+        for n_days_ago in CHANGE_LOOK_BACK:
+            acc: int = 0
+            for file in self.changed_files:
+                ps = subprocess.Popen(
+                    ["git", "log", f"--after='{n_days_ago} days ago'", "--format=oneline", f"{file}"],
+                    stdout=subprocess.PIPE,
+                )
+                try:
+                    acc += int(subprocess.check_output(["wc", "-l"], stdin=ps.stdout))
+                finally:
+                    ps.stdout.close()
+                    ps.wait()
+
+            modification_count.append(acc)
+        self.modification_count = modification_count
+
+    def _compute_file_extensions(self) -> None:
+        """
+        Computes the set of file extensions present in this code change
+        """
+        self.file_extensions = set([os.path.splitext(file)[1].replace(".", "") for file in self.changed_files])
+
+    def get_modifications_data_format(self) -> str:
+        """
+        Converts the modification_count (List[int]) to a condensed csv format.
+        e.g [1, 2, 3] becomes "1,2,3"
+        """
+        return ",".join((str(count) for count in self.modification_count))
+
+    def get_file_extensions_data_format(self) -> str:
+        """
+        Converts the file_extensions (Set[str]) to a condensed csv format.
+        e.g {'txt', 'md', 'py'} becomes 'txt,md,py'
+        """
+        return ",".join(self.file_extensions)
+
+
+@dataclass
+class TestResult:
+    """
+    Collects all data pertaining to the test cases by parsing the test result xml file (produced by
+    '$ py.test --junit-xml=junit-py39.xml'):
+    - test_result_data -> Maps the fully qualified test name <test_file_name>.<test_case> to an integer denoting whether
+      this test failed or not.
+    """
+
+    test_result_data: MutableMapping[str, int] = field(init=False)
+
+    def parse(self) -> None:
+        self.test_result_data = {}
+        path: str = "junit-py39.xml"
```

Comment:
We cannot hard-code this to `junit-py39.xml`. ISO4 runs on Python36, so for iso4 this will be `junit-py36.xml`.